src/common/harvest/harvester.js

```diff
@@ -0,0 +1,247 @@
+import { cleanURL } from '../url/clean-url'
+import { obj, param } from '../url/encode'
+import { warn } from '../util/console'
+import { stringify } from '../util/stringify'
+import { getSubmitMethod, xhr as xhrMethod, xhrFetch as fetchMethod } from '../util/submit-data'
+
+export class Harvester {
+  constructor (agentRef) {
+    this.agentRef = agentRef
+
+    const autoFeaturesInstruments = Object.values(agentRef.features).filter(feature => feature.auto) // features without autoStart won't be fully initialized until later
+    const importingFeatures = []
+    // Since the harvester is expected to be initialized by PVE Aggregate module, at such time, all the features should be awaiting import though featAggregate may not be available just yet.
+    for (const feature of autoFeaturesInstruments) {
+      if (!feature.onAggregateImported) warn(47, feature.featureName)
+      else importingFeatures.push(feature)
+    }
+    Promise.all(importingFeatures.map(instrument => instrument.onAggregateImported)).then(loadedSuccessfullyArr => {
+      // Double check that all aggregates have been initialized, successfully or not, before starting harvest schedule, which only queries the succesfully loaded ones.
+      const featuresToHarvest = importingFeatures.filter((instrumentInstance, index) => loadedSuccessfullyArr[index])
+      this.initializedAggregates = featuresToHarvest.map(instrumentInstance => instrumentInstance.featAggregate)
+      this.#startTimer(agentRef.init.harvest.interval)
+    })
+
+    subscribeToEOL(() => { // do one last harvest round or check
+      this.initializedAggregates?.forEach(aggregateInst => { // let all features wrap up things needed to do before ANY harvest in case there's last minute cross-feature data dependencies
+        if (typeof aggregateInst.harvestOpts.beforeUnload === 'function') aggregateInst.harvestOpts.beforeUnload()
+      })
+      this.initializedAggregates?.forEach(aggregateInst => this.triggerHarvestFor(aggregateInst, { isFinalHarvest: true }))
+      /* This callback should run in bubble phase, so that that CWV api, like "onLCP", is called before the final harvest so that emitted timings are part of last outgoing. */
+    }, false)
+
+    /* Flush all buffered data if session resets and give up retries. This should be synchronous to ensure that the correct `session` value is sent.
+      Since session-reset generates a new session ID and the ID is grabbed at send-time, any delays or retries would cause the payload to be sent under the wrong session ID. */
+    agentRef.ee.on(SESSION_EVENTS.RESET, () => this.initializedAggregates?.forEach(aggregateInst => this.triggerHarvestFor(aggregateInst, { forceNoRetry: true })))
+  }
+
+  #startTimer (harvestInterval) {
+    const onHarvestInterval = () => {
+      this.initializedAggregates.forEach(aggregateInst => this.triggerHarvestFor(aggregateInst))
+      setTimeout(onHarvestInterval, harvestInterval * 1000) // repeat in X seconds
```

Comment:
Could this mean if (in theory, for whatever reason) `send` took a long time, that we could be accumulating data for longer than the set harvest interval? 🤔